## Final Project Submission

Please fill out:
* Student name: HELLEN SAMUEL
* Student pace: full time
* Scheduled project review date/time: 
* Instructor name: NIKITA
* Blog post URL:


In [329]:
# Your code here - remember to use markdown cells for comments as well!
#import standard packages
import pandas as pd
import numpy as np
import seaborn as sns
import sqlite3
import matplotlib.pyplot as plt
%matplotlib inline

THE NUMBERS DATASET

In [330]:
#looking at the data set
budget = pd.read_csv('tn.movie_budgets.csv')
budget.head(10)

id  release_date                                        movie  \
0   1  Dec 18, 2009                                       Avatar   
1   2  May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2   3   Jun 7, 2019                                 Dark Phoenix   
3   4   May 1, 2015                      Avengers: Age of Ultron   
4   5  Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   
5   6  Dec 18, 2015         Star Wars Ep. VII: The Force Awakens   
6   7  Apr 27, 2018                       Avengers: Infinity War   
7   8  May 24, 2007   Pirates of the Caribbean: At World’s End   
8   9  Nov 17, 2017                               Justice League   
9  10   Nov 6, 2015                                      Spectre   

  production_budget domestic_gross worldwide_gross  
0      $425,000,000   $760,507,625  $2,776,345,279  
1      $410,600,000   $241,063,875  $1,045,663,875  
2      $350,000,000    $42,762,350    $149,762,350  
3      $330,600,000   $459,005,868  $1,403,013,963  
4      $317,000,000   $620,181,382  $1,316,721,747  
5      $306,000,000   $936,662,225  $2,053,311,220  
6      $300,000,000   $678,815,482  $2,048,134,200  
7      $300,000,000   $309,420,425    $963,420,425  
8      $300,000,000   $229,024,295    $655,945,209  
9      $300,000,000   $200,074,175    $879,620,923

In [331]:
#getting the information of the above data set
budget.info

<bound method DataFrame.info of       id  release_date                                        movie  \
0      1  Dec 18, 2009                                       Avatar   
1      2  May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2      3   Jun 7, 2019                                 Dark Phoenix   
3      4   May 1, 2015                      Avengers: Age of Ultron   
4      5  Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   
...   ..           ...                                          ...   
5777  78  Dec 31, 2018                                       Red 11   
5778  79   Apr 2, 1999                                    Following   
5779  80  Jul 13, 2005                Return to the Land of Wonders   
5780  81  Sep 29, 2015                         A Plague So Pleasant   
5781  82   Aug 5, 2005                            My Date With Drew   

     production_budget domestic_gross worldwide_gross  
0         $425,000,000   $760,507,625  $2,776,345,279  
1  

Changing the columns thatbcontain special characters and changing their type from object to int

In [332]:
#The values for production_budget, domestic_gross and worldwide_gross are strings, clean and convert to floats.
budget['production_budget'] = budget['production_budget'].str.replace('$','').str.replace(',','').astype('float')
budget['domestic_gross'] = budget['domestic_gross'].str.replace('$','').str.replace(',','').astype('float')
budget['worldwide_gross'] = budget['worldwide_gross'].str.replace('$','').str.replace(',','').astype('float')

In [333]:
#Changing the type of the columns domestic , production and worldwide from
#Objt to Int

budget['domestic_gross']= pd.to_numeric(budget['domestic_gross'],errors='coerce')
budget['production_budget']= pd.to_numeric(budget['production_budget'],errors='coerce')
budget['worldwide_gross']= pd.to_numeric(budget['worldwide_gross'],errors='coerce')


In [334]:
# Changing the realese date colunm data type from obj to datetime MM/DD/YY


budget['release_date']=pd.to_datetime(budget['release_date'])
budget['release_date']=budget['release_date'].dt.strftime('%m-%d-%Y')
budget.head()

id release_date                                        movie  \
0   1   12-18-2009                                       Avatar   
1   2   05-20-2011  Pirates of the Caribbean: On Stranger Tides   
2   3   06-07-2019                                 Dark Phoenix   
3   4   05-01-2015                      Avengers: Age of Ultron   
4   5   12-15-2017            Star Wars Ep. VIII: The Last Jedi   

   production_budget  domestic_gross  worldwide_gross  
0        425000000.0     760507625.0     2.776345e+09  
1        410600000.0     241063875.0     1.045664e+09  
2        350000000.0      42762350.0     1.497624e+08  
3        330600000.0     459005868.0     1.403014e+09  
4        317000000.0     620181382.0     1.316722e+09

In [335]:
#Take the converted financial columns and find the profit (worldwide_gross - production_budget) as well as the return on investment (roi).
budget['profit'] = budget['worldwide_gross'] - budget['production_budget']
budget['roi'] = (budget['profit']/budget['production_budget']) 
budget

id release_date                                        movie  \
0      1   12-18-2009                                       Avatar   
1      2   05-20-2011  Pirates of the Caribbean: On Stranger Tides   
2      3   06-07-2019                                 Dark Phoenix   
3      4   05-01-2015                      Avengers: Age of Ultron   
4      5   12-15-2017            Star Wars Ep. VIII: The Last Jedi   
...   ..          ...                                          ...   
5777  78   12-31-2018                                       Red 11   
5778  79   04-02-1999                                    Following   
5779  80   07-13-2005                Return to the Land of Wonders   
5780  81   09-29-2015                         A Plague So Pleasant   
5781  82   08-05-2005                            My Date With Drew   

      production_budget  domestic_gross  worldwide_gross        profit  \
0           425000000.0     760507625.0     2.776345e+09  2.351345e+09   
1           410600000.0     241063875.0     1.045664e+09  6.350639e+08   
2           350000000.0      42762350.0     1.497624e+08 -2.002376e+08   
3           330600000.0     459005868.0     1.403014e+09  1.072414e+09   
4           317000000.0     620181382.0     1.316722e+09  9.997217e+08   
...                 ...             ...              ...           ...   
5777             7000.0             0.0     0.000000e+00 -7.000000e+03   
5778             6000.0         48482.0     2.404950e+05  2.344950e+05   
5779             5000.0          1338.0     1.338000e+03 -3.662000e+03   
5780             1400.0             0.0     0.000000e+00 -1.400000e+03   
5781             1100.0        181041.0     1.810410e+05  1.799410e+05   

             roi  
0       5.532577  
1       1.546673  
2      -0.572108  
3       3.243841  
4       3.153696  
...          ...  
5777   -1.000000  
5778   39.082500  
5779   -0.732400  
5780   -1.000000  
5781  163.582727  

[5782 rows x 8 columns]

In [336]:
#looking at the changes made
budget.head()

id release_date                                        movie  \
0   1   12-18-2009                                       Avatar   
1   2   05-20-2011  Pirates of the Caribbean: On Stranger Tides   
2   3   06-07-2019                                 Dark Phoenix   
3   4   05-01-2015                      Avengers: Age of Ultron   
4   5   12-15-2017            Star Wars Ep. VIII: The Last Jedi   

   production_budget  domestic_gross  worldwide_gross        profit       roi  
0        425000000.0     760507625.0     2.776345e+09  2.351345e+09  5.532577  
1        410600000.0     241063875.0     1.045664e+09  6.350639e+08  1.546673  
2        350000000.0      42762350.0     1.497624e+08 -2.002376e+08 -0.572108  
3        330600000.0     459005868.0     1.403014e+09  1.072414e+09  3.243841  
4        317000000.0     620181382.0     1.316722e+09  9.997217e+08  3.153696

In [337]:
budget.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5782 non-null   int64  
 1   release_date       5782 non-null   object 
 2   movie              5782 non-null   object 
 3   production_budget  5782 non-null   float64
 4   domestic_gross     5782 non-null   float64
 5   worldwide_gross    5782 non-null   float64
 6   profit             5782 non-null   float64
 7   roi                5782 non-null   float64
dtypes: float64(5), int64(1), object(2)
memory usage: 361.5+ KB


In [338]:
#checking if our dataframe has any missing values
budget.isna().sum()

id                   0
release_date         0
movie                0
production_budget    0
domestic_gross       0
worldwide_gross      0
profit               0
roi                  0
dtype: int64

The dataframe is clean as it doesn't have any missing values

IMBD DATASET

In [339]:
import zipfile

with zipfile.ZipFile('zippedData/im.db.zip') as my_zip:
    zipfile.ZipFile.extractall(my_zip, path='ZippedData')

#open up a connection
conn = sqlite3.connect('zippedData/im.db')
#initialize connection
cursor = conn.cursor()

In [340]:
imdbrating =pd.read_sql( """
SELECT name 
FROM sqlite_master 
WHERE type ='table' AND name NOT LIKE 'sqlite_%';
""", conn)

imdbrating



name
0   movie_basics
1      directors
2      known_for
3     movie_akas
4  movie_ratings
5        persons
6     principals
7        writers

In [341]:
#checking the movie rating
imdbrating = pd.read_sql("""SELECT * FROM movie_ratings""",conn)
imdbrating.head()


movie_id  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [342]:
#checking the information in the dataframe
imdbrating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   movie_id       73856 non-null  object 
 1   averagerating  73856 non-null  float64
 2   numvotes       73856 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 1.7+ MB


In [343]:
#checking if our data has any missing values
imdbrating.isna().sum()

movie_id         0
averagerating    0
numvotes         0
dtype: int64

the movie_ratings data set is clean now we look at movie_basics

In [344]:
imdbtitle = pd.read_sql("""SELECT * FROM movie_basics""",conn)
imdbtitle.head()


movie_id                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [345]:
#checking the dataframe information
imdbtitle.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   movie_id         146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


correcting the missing values in our dataframe

In [346]:
#dropping the original_title and changing primary_title to movie
imdbtitle = imdbtitle.drop('original_title', 1)
imdbtitle.rename(columns={'primary_title': 'movie'}, inplace=True)

imdbtitle.head(25)


movie_id                                              movie  start_year  \
0   tt0063540                                          Sunghursh        2013   
1   tt0066787                    One Day Before the Rainy Season        2019   
2   tt0069049                         The Other Side of the Wind        2018   
3   tt0069204                                    Sabse Bada Sukh        2018   
4   tt0100275                           The Wandering Soap Opera        2017   
5   tt0111414                                        A Thin Life        2018   
6   tt0112502                                            Bigfoot        2017   
7   tt0137204                                    Joe Finds Grace        2017   
8   tt0139613                                         O Silêncio        2012   
9   tt0144449                              Nema aviona za Zagreb        2012   
10  tt0146592                                        Pál Adrienn        2010   
11  tt0154039                               So Much for Justice!        2010   
12  tt0159369                 Cooper and Hemingway: The True Gen        2013   
13  tt0162942                       Children of the Green Dragon        2010   
14  tt0170651                               T.G.M. - osvoboditel        2018   
15  tt0176694                                 The Tragedy of Man        2011   
16  tt0187902  How Huang Fei-hong Rescued the Orphan from the...        2011   
17  tt0192528                                      Heaven & Hell        2018   
18  tt0230212                                  The Final Journey        2010   
19  tt0247643                   Los pájaros se van con la muerte        2011   
20  tt0249516                                         Foodfight!        2012   
21  tt0250404                                          Godfather        2012   
22  tt0253093                                       Gangavataran        2018   
23  tt0255820                                  Return to Babylon        2013   
24  tt0262218                                  Akakis mogzauroba        2012   

    runtime_minutes                      genres  
0             175.0          Action,Crime,Drama  
1             114.0             Biography,Drama  
2             122.0                       Drama  
3               NaN                Comedy,Drama  
4              80.0        Comedy,Drama,Fantasy  
5              75.0                      Comedy  
6               NaN             Horror,Thriller  
7              83.0  Adventure,Animation,Comedy  
8               NaN         Documentary,History  
9              82.0                   Biography  
10            136.0                       Drama  
11            100.0                     History  
12            180.0                 Documentary  
13             89.0                       Drama  
14             60.0                 Documentary  
15            160.0     Animation,Drama,History  
16              NaN                        None  
17            104.0                       Drama  
18            120.0                       Drama  
19            110.0               Drama,Mystery  
20             91.0     Action,Animation,Comedy  
21              NaN                 Crime,Drama  
22            134.0                        None  
23             75.0      Biography,Comedy,Drama  
24             44.0                 Documentary

In [347]:
imdbtitle['movie'].duplicated().sum()

10073

In [348]:
imdbtitle.loc[imdbtitle['movie'].str.contains('Avatar',case=False)]

movie_id                               movie  start_year  \
22       tt0253093                        Gangavataran        2018   
1795    tt10168040                     My Avatar Horse        2012   
4315    tt10403262                     Xbox Avatar Fun        2019   
11392    tt1630029                            Avatar 2        2021   
14612    tt1720145  Les avatars de la mort d'Empédocle        2010   
16337    tt1757678                            Avatar 3        2023   
17060    tt1775309                              Avatar        2011   
20368    tt1854506                  Aliens vs. Avatars        2011   
34245    tt2271191              Avatars as Prostitutes        2012   
51456    tt3053768                        Koormavatara        2011   
52213    tt3095356                            Avatar 4        2025   
68178    tt3879186         Avatarii: Birth of a Fandom        2014   
77990    tt4472992                   Avatarachi Goshta        2014   
82231    tt4727514                     The Last Avatar        2014   
88870    tt5141528                            Avataran        2016   
96592    tt5637536                            Avatar 5        2027   
112823   tt6620864                  The Avatar Project        2017   

        runtime_minutes                       genres  
22                134.0                         None  
1795                NaN                        Drama  
4315                NaN                       Action  
11392               NaN     Action,Adventure,Fantasy  
14612              53.0                  Documentary  
16337               NaN       Action,Adventure,Drama  
17060              93.0                       Horror  
20368              80.0                Horror,Sci-Fi  
34245              51.0  Animation,Documentary,Drama  
51456               NaN                        Drama  
52213               NaN     Action,Adventure,Fantasy  
68178               NaN                  Documentary  
77990             118.0                        Drama  
82231              90.0                        Drama  
88870               NaN                       Sci-Fi  
96592               NaN     Action,Adventure,Fantasy  
112823             66.0                     Thriller

dropping the duplicate rows that we have

In [349]:
imdbtitle.drop_duplicates(subset="movie", keep= False, inplace=True)

In [350]:
imdbtitle['movie'].duplicated().sum()

0

The duplicate rows have been dropped and now we look at the columns

In [351]:
#looking at our columns
imdbtitle.isna()

movie_id  movie  start_year  runtime_minutes  genres
0          False  False       False            False   False
1          False  False       False            False   False
2          False  False       False            False   False
3          False  False       False             True   False
4          False  False       False            False   False
...          ...    ...         ...              ...     ...
146139     False  False       False            False   False
146140     False  False       False             True   False
146141     False  False       False             True   False
146142     False  False       False            False    True
146143     False  False       False             True   False

[129889 rows x 5 columns]

In [352]:
imdbtitle.isnull().sum()

movie_id               0
movie                  0
start_year             0
runtime_minutes    28397
genres              4709
dtype: int64

Here we have some missing data in runtime_minutes and genre

In [353]:
#exploring the genre column
imdbtitle['genres'].describe()

count          125180
unique           1058
top       Documentary
freq            30177
Name: genres, dtype: object

In [354]:
imdbt = (imdbtitle['genres'].isnull().sum()
         /len(imdbtitle))*100
print('The percentage of missing genres is {}%'.format(round(imdbt, 2)))

The percentage of missing genres is 3.63%


3.63% of data is missing from our genre column

In [355]:
#replacing the missing values with the most common genre in our data frame which is the Documentary
imdbtitle['genres'] = imdbtitle['genres'].fillna('Documentary')

In [356]:
#confirming if the code worked
imdbtitle.isnull().sum()

movie_id               0
movie                  0
start_year             0
runtime_minutes    28397
genres                 0
dtype: int64

The missing data in genre column has been corrected

In [357]:
# Exploring the runtime column
imdbr = (imdbtitle['runtime_minutes'].isnull().sum()
         /len(imdbtitle))*100
print('The percentage of missing runtime data is {}%'.format(round(imdbr, 2)))

The percentage of missing runtime data is 21.86%


The missing data percentage is 21.86

In [358]:
imdbtitle['runtime_minutes'].describe()

count    101492.000000
mean         86.037924
std         176.204615
min           1.000000
25%          69.000000
50%          86.000000
75%          99.000000
max       51420.000000
Name: runtime_minutes, dtype: float64

In [359]:
imdbtitle.median()

start_year         2015.0
runtime_minutes      86.0
dtype: float64

In [360]:
#Replacing the NaN values  of our dataset with the median column
imdbtitle['runtime_minutes'] = imdbtitle['runtime_minutes'].fillna(imdbtitle['runtime_minutes'].median())

In [361]:
#Confirming our code
imdbtitle.isnull().sum()

movie_id           0
movie              0
start_year         0
runtime_minutes    0
genres             0
dtype: int64

The missing values in runtime_minutes are now okay

In [362]:
#let's check our Data Frame
imdbtitle.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 129889 entries, 0 to 146143
Data columns (total 5 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   movie_id         129889 non-null  object 
 1   movie            129889 non-null  object 
 2   start_year       129889 non-null  int64  
 3   runtime_minutes  129889 non-null  float64
 4   genres           129889 non-null  object 
dtypes: float64(1), int64(1), object(3)
memory usage: 5.9+ MB


In [363]:
#previewing the persons table
q3 = pd.read_sql("""
SELECT *
FROM persons
;
""",conn)
q3.head()


person_id       primary_name  birth_year  death_year  \
0  nm0061671  Mary Ellen Bauder         NaN         NaN   
1  nm0061865       Joseph Bauer         NaN         NaN   
2  nm0062070         Bruce Baum         NaN         NaN   
3  nm0062195       Axel Baumann         NaN         NaN   
4  nm0062798        Pete Baxter         NaN         NaN   

                                 primary_profession  
0         miscellaneous,production_manager,producer  
1        composer,music_department,sound_department  
2                        miscellaneous,actor,writer  
3  camera_department,cinematographer,art_department  
4  production_designer,art_department,set_decorator

MERGING DATAFRAMES

In [364]:
# let's merge our dataframes, lets start first with our IMDB dataframes , imdbtitle and imdb rating
imdbdata = pd.merge(left=imdbtitle, right=imdbrating, on="movie_id")

In [366]:
imdbdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 64842 entries, 0 to 64841
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   movie_id         64842 non-null  object 
 1   movie            64842 non-null  object 
 2   start_year       64842 non-null  int64  
 3   runtime_minutes  64842 non-null  float64
 4   genres           64842 non-null  object 
 5   averagerating    64842 non-null  float64
 6   numvotes         64842 non-null  int64  
dtypes: float64(2), int64(2), object(3)
memory usage: 4.0+ MB


In [367]:
imdbdata.head()

movie_id                            movie  start_year  runtime_minutes  \
0  tt0063540                        Sunghursh        2013            175.0   
1  tt0066787  One Day Before the Rainy Season        2019            114.0   
2  tt0069049       The Other Side of the Wind        2018            122.0   
3  tt0069204                  Sabse Bada Sukh        2018             86.0   
4  tt0100275         The Wandering Soap Opera        2017             80.0   

                 genres  averagerating  numvotes  
0    Action,Crime,Drama            7.0        77  
1       Biography,Drama            7.2        43  
2                 Drama            6.9      4517  
3          Comedy,Drama            6.1        13  
4  Comedy,Drama,Fantasy            6.5       119